feat(compile): compile taproot descriptor with randomized unspendable internal key

va-an opened the pull request:
### Description

Part of #218

Implements randomization of the unspendable internal key for taproot descriptors. This is the first part of #218, which consists of two parts:
1. **This PR**: Randomize unspendable internal key for taproot descriptor
2. **Follow-up**: Add verification command to ensure internal key is derived from NUMS key

Split into separate PRs for easier review and iteration, and to allow independent discussion of the verification command implementation, as one of the possible approaches could introduce breaking changes.

### Notes to the reviewers
The `compile` command now returns an additional `r` field for taproot descriptors (`-t tr`), containing the randomly generated internal key. Each compilation will produce a different internal key instead of using a fixed NUMS key.

Example output for taproot (first execution):
```
-> % bdk-cli compile "pk(A)" -t tr
{
  "descriptor": "tr(2dd09dd0355f4b2d5a4886de599786f3c0211652373221c87aba1cd1f7f1e593,pk(A))#anvu48aj",
  "r": "275a58827bd0ad459d6f92e083ddc3d99a03076155691680eb8f3b06380cdcfd"
}
```

Same descriptor compiled again produces different `r` and internal key:
```
-> % bdk-cli compile "pk(A)" -t tr
{
  "descriptor": "tr(801078f69dae7d95631723d4d13e6c32911633d227dcfc24c6b7e32e1e533e6c,pk(A))#f79rr82j",
  "r": "5e3ac63bb20d6a4bfff645279cc63a7472e18066da8826b13cbcb23aecb5c401"
}
```

Other descriptor types remain unchanged:
```
-> % bdk-cli compile "pk(A)" -t sh
{
  "descriptor": "sh(pk(A))#k80zhe7s"
}
```

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk-cli/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

tvpeter commented on Cargo.toml:
```diff
@@ -54,3 +54,6 @@ verify = []
 compiler = []
+
+[dev-dependencies]
+shlex = "1.3.0"
```
`shlex` is an optional dependency, so instead of re-installing it under dev dependencies, it's better to make it a non-optional dependency.

va-an replied:
`shlex` needs to remain both optional and in dev-dependencies:
  - Optional: it's only used by the `repl` feature, so other builds don't need it
  - Dev-dependencies: CI runs `cargo test --no-default-features` which excludes `repl`, but tests still need `shlex` to compile

  Making it non-optional would unnecessarily add the dependency to all builds.

  Alternatively, we could accept the dependency overhead but simplify the code and CI in return?

tvpeter commented on src/handlers.rs:
```diff
@@ -997,50 +1003,63 @@ pub(crate) fn handle_compile_subcommand(
 
-            let xonly_public_key = XOnlyPublicKey::from_str(NUMS_UNSPENDABLE_KEY_HEX)
-                .map_err(|e| Error::Generic(format!("Invalid NUMS key: {e}")))?;
+            let internal_key_point = nums_point.combine(&r_point)?;
```
Have you considered the `add_exp_tweak` method? I think it might be better suited for this use case because it knows one of the factors is a Generator point, whereas `combine` is for adding arbitrary public points.

va-an replied:
Good point, thanks! Done.

tvpeter reviewed (changes requested):
Thanks for working on this @va-an 
I have left a few comments.

va-an commented on src/error.rs:
```diff
@@ -45,6 +45,9 @@ pub enum BDKCliError {
 
+    #[error("Miniscript compiler error: {0}")]
+    MiniscriptCompilerError(#[from] bdk_wallet::miniscript::policy::compiler::CompilerError),
+
```
I added this error only for creating policies in a more concise way:
https://github.com/bitcoindevkit/bdk-cli/pull/225/files#diff-148e623ad4cdaab49b81901460099f4efe7900e8c57e4f7a0ff6778589be01f8L896-R905

Which way do you prefer more?
1) remove error, restore using err_map for policies
2) keep error and create policies without err_map

110CodingP reviewed (commented):
Refactoring the tests to use`std::process::Command` instead of calling the internal function `handle_compile_subcommand` make the tests vulnerable to the state of environment variables and makes debugging harder in my opinion.  But I have no strong disagreement to the approach in the PR.

va-an commented:
@110CodingP I decided to add the code for removing env vars in tests, as we discussed, so we don't forget to do this in future PRs. 

I verified locally that this works as expected — the command process will not receive these env vars from the parent process.
```rust
use std::{env, process::Command};

fn main() {
    unsafe {
        env::set_var("FOO", "1");
        env::set_var("BAR", "2");
        env::set_var("BAZ", "3");
    }

    let cmd = Command::new("printenv")
        .env_remove("FOO")
        .env_remove("BAR")
        .output()
        .unwrap();

    let stdout = String::from_utf8_lossy(&cmd.stdout);

    assert!(!stdout.contains("FOO"), "FOO should be removed");
    assert!(!stdout.contains("BAR"), "BAR should be removed");
    assert!(stdout.contains("BAZ"), "BAZ should still be present");

    println!("it's all good man");
}
```

Please take a look — https://github.com/bitcoindevkit/bdk-cli/pull/225/commits/afc76aa8d44eb609548d0923c59cd55f6785a6bd.

tvpeter reviewed (commented):
Thank you for updating the PR @va-an 

Once you address the little comments, I will do a final test.

va-an commented on tests/compile.rs:
```diff
@@ -0,0 +1,78 @@
+    let output = Command::new("cargo")
+        .args(args)
+        .env_remove("NETWORK")
+        .env_remove("DATADIR")
```
You're right, `DATADIR` isn't used in compile command. 
I'm planning to reuse this command execution code in other tests, so I preemptively cleared all env vars.

va-an commented on src/handlers.rs:
```diff
@@ -893,50 +899,61 @@ pub(crate) fn handle_compile_subcommand(
-        .table()
-        .display()
-        .map_err(|e| Error::Generic(e.to_string()))?;
+        let mut rows = vec![vec!["Descriptor".cell().bold(true), descriptor.cell()]];
```
Done! I also used `shorten()` for r.
I extracted shorten description into a variable to shorten only the internal key and leave the rest of the descriptor as is.

without pretty:
```bash
-> % cargo run --all-features -- compile "pk(ABC)" -t tr
{
  "descriptor": "tr(1985c2a86e01bbcdb1b5806422531a73c77e568ecfdd6d2863c158a76628ea50,pk(ABC))#6qsjh5q3",
  "r": "e38caa40d0db7cae7037336c982278a7fc6b770e7a80f8bbce0d79976a14d5e1"
}
```

with pretty:
```rust
-> % cargo run --all-features -- compile "pk(ABC)" -t tr --pretty
+------------+----------------------------------+
| Descriptor | tr(ccb4...b1d0,pk(ABC))#qrk923py |
+------------+----------------------------------+
| r          | a129...7a14                      |
+------------+----------------------------------+
```

tvpeter commented on src/handlers.rs:
```diff
@@ -893,50 +899,70 @@ pub(crate) fn handle_compile_subcommand(
-        .map_err(|e| Error::Generic(e.to_string()))?;
+        let descriptor = shorten_descriptor.unwrap_or(descriptor);
+
+        let mut rows = vec![vec!["Descriptor".cell().bold(true), descriptor.cell()]];
```
```suggestion
let mut rows = vec![vec!["Descriptor".cell().bold(true), shorten(descriptor, 32, 29).cell()]];
```

you can remove: 908,  931-934, 946

you can also revert 951; it should size most screens comfortably.

va-an replied:
Ok, I’ll add these changes when I do the squash.

tvpeter replied:
Yeah, I noticed this bug and fixed it in #203 [here](https://github.com/bitcoindevkit/bdk-cli/pull/203/changes#diff-23af5356f6001cd3993cd3c801fb7716ea02d1e504081b9fb569332db6107e80R367-R370). We may have to wait for the PR to get merged.

tvpeter reviewed (commented):
Hi @va-an, 

Could you please squash your commits into two or three commits and fix the shortening approach?

Aside from that, it looks good to go. Thank you.

notmandatory reviewed (commented):
Overall the code looks good, my only concern is with running shell commands as part of the testing, which I'd like to see removed.

Ultimately I'd like to figure out how to do the TR optimizatoin for the `pk(A)` policy so that it creates the descriptor `tr(A)` but I don't know if/how to do it or if something would need to be changed upstream.

va-an committed:
chore(deps): make shlex a non-optional dependency

Although not directly related to this PR's changes, during review
we agreed to make shlex non-optional since it's used by the default
`repl` feature and the package is under 20 KiB.

See: https://github.com/bitcoindevkit/bdk-cli/pull/225#discussion_r2671741961

va-an committed:
feat(compile): randomize unspendable internal key for taproot descriptor

Instead of using a fixed NUMS key as the internal key for taproot
descriptors, generate a randomized unspendable key (H + rG) for each
compilation. This improves privacy by preventing observers from
determining whether key path spending is disabled.

The randomness factor `r` is included in the output so the user can
verify that the internal key is derived from the NUMS point.

Also applies `shorten()` globally in pretty mode and uses `?` operator
via dedicated error variants instead of `map_err`.

Tests verify descriptor structure and `r` field uniqueness across
randomized compilations.

va-an committed:
refactor(compile): lazy compilation and use pipe for readability

Move miniscript compilation inside match arms to avoid compiling
for all script contexts when only one is needed. Use `tap::Pipe`
for concise method chaining in sh/wsh/sh-wsh branches.

va-an commented on src/handlers.rs:
```diff
@@ -1014,50 +1022,61 @@ pub(crate) fn handle_compile_subcommand(
 
-            let xonly_public_key = XOnlyPublicKey::from_str(NUMS_UNSPENDABLE_KEY_HEX)
-                .map_err(|e| Error::Generic(format!("Invalid NUMS key: {e}")))?;
+            let tree = TapTree::Leaf(Arc::new(policy.compile::<Tap>()?));
```
I'm planning to replace `compile()` with `compile_tr()` for the `tr` match branch, and moving the compile step inside the match is preparation for that.

See: https://github.com/bitcoindevkit/bdk-cli/issues/244

va-an committed:
docs(contributing): update CONTRIBUTING.md

- Remove MSRV mention as the project no longer enforces it
- Add Conventional Commits and GPG signing links to align with other repos

vadim-anfv commented on Cargo.toml:
```diff
@@ -12,7 +12,7 @@ readme = "README.md"
 
 [dependencies]
-bdk_wallet = { version = "2.1.0", features = ["rusqlite", "keys-bip39", "compiler", "std"] }
+bdk_wallet = { version = "=2.1.0", features = ["rusqlite", "keys-bip39", "compiler", "std"] }
```
See #254

tvpeter replied:
Pls revert because of [this](https://github.com/bitcoindevkit/bdk-cli/issues/254#issuecomment-4161840037)

va-an committed:
docs: add pre-push recipe description to README

va-an commented:
Hey @tvpeter, is there anything else you'd like me to do on my end? Should I squash the follow-up commits down again (a few have piled up since the last squash)?

Thanks!

tvpeter commented:
> Hey @tvpeter, is there anything else you'd like me to do on my end? Should I squash the follow-up commits down again (a few have piled up since the last squash)?

Hi @va-an, apologies for the delay. I will give it a final check shortly.

tvpeter reviewed (commented):
hi @va-an , I have reviewed again and left one minor nit below.

Also, kindly remove these [lines](https://github.com/bitcoindevkit/bdk-cli/blob/f80d92a549ea648b65595a445cf1c6c187b7b337/src/handlers.rs#L1739-L1745) because the gating does not apply to compiler feature tests, and there is no way to run the compile feature tests except `all-features` are compiled.

Please, once you have effected these changes, tag me to approve.

Thank you.

va-an committed:
test(handlers): drop redundant cfg gate on test module

tvpeter reviewed (approved):
Thank you @va-an.

While it is not mandatory, you can squash 05a248cbd0bde6ecd448369067f71a5518bd743a, 
1ed41d21ec0c9e898e231b958a787f96dcf9fdba and bc8135d8d3871c2087b98187caab40cae58858d8.

Also, please revert the bdK_wallet version to 2.1, as some features in versions above 2.1 have not yet been implemented in the CLI.

Thank you
